### PR TITLE
Enforce that contrast limits must be increasing

### LINKED ISSUE #4158: Setting contrast_limits to [1,1] causes hard to debug error
## 🐛 Bug

Hi all,

I just debugged an issue deep in napari and it turned out, I was unintentionally setting contrast_limits to [1,1]. If min and max of contrast limits are equal, this causes an error and it took me quite some time to understand what's the reason. I'm suggesting to catch this situation early with a clear error message. 

## To Reproduce

From python (I did in a juptyer notebook), call this:
```
import napari
import numpy as np

viewer = napari.Viewer()
image = np.asarray([[0,1], [0, np.inf]])
viewer.add_image(image, contrast_limits=[1,1])
```

This error happens:

```
File ~\miniconda3\envs\bio_39\lib\site-packages\napari\_qt\layer_controls\qt_image_controls_base.py:287, in range_to_decimals(range_, dtype)
[...]
OverflowError: cannot convert float infinity to integer
```

<details>
<summary markdown="span">Detailed log</summary>
<pre>
divide by zero encountered in log10
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Input In [3], in <module>
      1 image = np.asarray([[0,1], [0, np.inf]])
----> 3 viewer.add_image(image, contrast_limits=[1,1])

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\components\viewer_model.py:732, in ViewerModel.add_image(self, data, channel_axis, rgb, colormap, contrast_limits, gamma, interpolation, rendering, iso_threshold, attenuation, name, metadata, scale, translate, rotate, shear, affine, opacity, blending, visible, multiscale, cache, experimental_slicing_plane, experimental_clipping_planes)
    724             raise TypeError(
    725                 trans._(
    726                     "Received sequence for argument '{argument}', did you mean to specify a 'channel_axis'? ",
   (...)
    729                 )
    730             )
    731     layer = Image(data, **kwargs)
--> 732     self.layers.append(layer)
    734     return layer
    735 else:

File ~\miniconda3\envs\bio_39\lib\_collections_abc.py:1073, in MutableSequence.append(self, value)
   1071 def append(self, value):
   1072     'S.append(value) -- append value to the end of the sequence'
-> 1073     self.insert(len(self), value)

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\components\layerlist.py:90, in LayerList.insert(self, index, value)
     88 new_layer = self._type_check(value)
     89 new_layer.name = self._coerce_name(new_layer.name)
---> 90 super().insert(index, new_layer)

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\utils\events\containers\_selectable_list.py:65, in SelectableEventedList.insert(self, index, value)
     64 def insert(self, index: int, value: _T):
---> 65     super().insert(index, value)
     66     if self._activate_on_insert:
     67         # Make layer selected and unselect all others
     68         self.selection.active = value

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\utils\events\containers\_evented_list.py:179, in EventedList.insert(self, index, value)
    177 self.events.inserting(index=index)
    178 super().insert(index, value)
--> 179 self.events.inserted(index=index, value=value)
    180 self._connect_child_emitters(value)

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\utils\events\event.py:715, in EventEmitter.__call__(self, *args, **kwargs)
    712     self._block_counter.update([cb])
    713     continue
--> 715 self._invoke_callback(cb, event if pass_event else None)
    716 if event.blocked:
    717     break

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\utils\events\event.py:752, in EventEmitter._invoke_callback(self, cb, event)
    750     self.disconnect(cb)
    751     return
--> 752 _handle_exception(
    753     self.ignore_callback_errors,
    754     self.print_callback_errors,
    755     self,
    756     cb_event=(cb, event),
    757 )

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\utils\events\event.py:739, in EventEmitter._invoke_callback(self, cb, event)
    737 try:
    738     if event is not None:
--> 739         cb(event)
    740     else:
    741         cb()

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\_qt\layer_controls\qt_layer_controls_container.py:128, in QtLayerControlsContainer._add(self, event)
    120 """Add the controls target layer to the list of control widgets.
    121 
    122 Parameters
   (...)
    125     Event with the target layer at `event.value`.
    126 """
    127 layer = event.value
--> 128 controls = create_qt_layer_controls(layer)
    129 self.addWidget(controls)
    130 self.widgets[layer] = controls

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\_qt\layer_controls\qt_layer_controls_container.py:67, in create_qt_layer_controls(layer)
     65 candidates.sort(key=lambda layer_type: layer_cls.mro().index(layer_type))
     66 controls = layer_to_controls[candidates[0]]
---> 67 return controls(layer)

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\_qt\layer_controls\qt_image_controls.py:46, in QtImageControls.__init__(self, layer)
     45 def __init__(self, layer):
---> 46     super().__init__(layer)
     48     self.layer.events.interpolation.connect(self._on_interpolation_change)
     49     self.layer.events.rendering.connect(self._on_rendering_change)

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\_qt\layer_controls\qt_image_controls_base.py:92, in QtBaseImageControls.__init__(self, layer)
     90 # Create contrast_limits slider
     91 self.contrastLimitsSlider = _QDoubleRangeSlider(Qt.Horizontal, self)
---> 92 decimals = range_to_decimals(
     93     self.layer.contrast_limits_range, self.layer.dtype
     94 )
     95 self.contrastLimitsSlider.setRange(*self.layer.contrast_limits_range)
     96 self.contrastLimitsSlider.setSingleStep(10 ** -decimals)

File ~\miniconda3\envs\bio_39\lib\site-packages\napari\_qt\layer_controls\qt_image_controls_base.py:287, in range_to_decimals(range_, dtype)
    279 else:
    280     # scale precision with the log of the data range order of magnitude
    281     # eg.   0 - 1   (0 order of mag)  -> 3 decimal places
   (...)
    284     #       ≥ 3 orders of mag -> no decimals
    285     # no more than 64 decimals
    286     d_range = np.subtract(*range_[::-1])
--> 287     return min(64, max(int(3 - np.log10(d_range)), 0))

OverflowError: cannot convert float infinity to integer
</pre>
</details>

## Expected behavior

Either no error happens at all, or a clear warning message tells that contrast_limits must be two different numbers. I'd be happy to implement the second case , e.g. [here](https://github.com/napari/napari/blob/main/napari/layers/image/image.py#L329), and send a PR if there is concensus that this is the right place and strategy.

## Environment

 - Please copy and paste the information at napari info option in help menubar here:
```
napari: 0.4.14
Platform: Windows-10-10.0.19042-SP0
Python: 3.9.10 | packaged by conda-forge | (main, Feb 1 2022, 21:22:07) [MSC v.1929 64 bit (AMD64)]
Qt: 5.12.9
PyQt5: 5.12.3
NumPy: 1.21.5
SciPy: 1.8.0
Dask: 2022.02.0
VisPy: 0.9.6

OpenGL:
- GL version: 4.6.14761 Compatibility Profile Context 21.30.02.01 30.0.13002.1001
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1200, scale 1.0

Plugins:
- PlatyMatch: 0.0.3
- RedLionfish: 0.6
- aicsimageio-in-memory: 0.5.0
- aicsimageio-out-of-memory: 0.5.0
- animation: 0.0.2
- cellpose-napari: 0.1.4
- clEsperanto: 0.15.4
- console: 0.0.4
- devbio_napari: 0.4.0
- napari-3d-ortho-viewer: 0.0.1
- napari-accelerated-pixel-and-object-classification: 0.6.6
- napari-allencell-segmenter: 1.1.4
- napari-brightness-contrast: 0.1.6
- napari-clusters-plotter: 0.2.2
- napari-crop: 0.1.5
- napari-curtain: 0.1.1
- napari-folder-browser: 0.1.3
- napari-itk-io: 0.1.0
- napari-ome-zarr: 0.1.0
- napari-plot-profile: 0.2.1
- napari-plugin-search: 0.1.3
- napari-segment-blobs-and-things-with-membranes: 0.2.17
- napari-simpleitk-image-processing: 0.2.1
- napari-tabu: 0.1.5
- napari-time-slicer: 0.4.3
- napari-tools-menu: 0.1.10
- napari-workflow-optimizer: 0.1.3
- napari_skimage_regionprops1: 0.2.9
- napari_skimage_regionprops2: 0.2.9
- ome-types: 0.2.10
- ome_zarr: 0.0.18
- scikit-image: 0.4.14
- svg: 0.1.5
```
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by kcpevey**
# Description

Ensures that contrast limits are increasing

This is a continuation of the work started by @haesleinhuepf in PR #4159

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

Closes #4158

# How has this been tested?

- [x] example: all tests pass with my change 

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by kcpevey** [comment]
I could use a bit of help with the CI failure. I don't understand the difference between `contrast_limits` and `contrast_limits_range`. They aren't well defined in the doc string and the code around them isn't very clean - are they just the same thing??

```
    contrast_limits : list (2,) of float
        Color limits to be used for determining the colormap bounds for
        luminance images. If the image is rgb the contrast_limits is ignored.
    contrast_limits_range : list (2,) of float
        Range for the color limits for luminance images. If the image is
        rgb the contrast_limits_range is ignored.
```

**Comment by psobolewskiPhD** [comment]
> I could use a bit of help with the CI failure. I don't understand the difference between `contrast_limits` and `contrast_limits_range`. They aren't well defined in the doc string and the code around them isn't very clean - are they just the same thing??
> 
> ```
>     contrast_limits : list (2,) of float
>         Color limits to be used for determining the colormap bounds for
>         luminance images. If the image is rgb the contrast_limits is ignored.
>     contrast_limits_range : list (2,) of float
>         Range for the color limits for luminance images. If the image is
>         rgb the contrast_limits_range is ignored.
> ```

Based on using napari, if have, say a greyscale 0-255 image and you use:
`viewer.layers[0].contrast_limits = [50, 150]`
Then you will change the set contrast limits to 50 (lower) 150 (upper) for display, out of a possible range of 0-255. Equivalent to moving the slider in the UI and making the positions of the end stops closer together.
image

Meanwhile, if you use:
`viewer.layers[0].contrast_limits_range = [50, 150]`
The visual effect on the canvas/image will be the same, but the slider will not permit setting outside that range: the slider end stops will still be at maximum. To change outside the range in the UI, one needs to right click the slider and change the range.
image
`viewer.layers[0].contrast_limits` overrides any `_range` setting and sets a new range if outside. So: `viewer.layers[0].contrast_limits = [0, 150]`:
image
```
viewer.layers[0].contrast_limits_range
Out[4]: [0, 150]
```

So the `contrast_limits` are always *within* the `contrast_limits_range`. If set outside the range, the range is reset to the `contrast_limits` value that is minimal/maximal. 
The `contrast_limits_range` is what determines the UI display of the contrast limits slider.

**Comment by brisvag** [comment]
Is there any reason to keep the `range` on the napari/core side rather than the qt/gui side?

**Comment by kcpevey** [comment]
Thanks @psobolewskiPhD ! That was super helpful. `contrast_limits_range` is range extent available on the widget and `contrast_limits` is the visible range (the values the widget is set at)

**Comment by kcpevey** [comment]
What is the expected behavior of `contrast_limits_range` is set outside of `contrast_limits`? 

An existing test expects this:
```
    layer.contrast_limits_range = [0, 100]
    layer.contrast_limits = [20, 40]
    layer.contrast_limits_range = [60, 100]
    assert layer.contrast_limits == [60, 60]
```

But the new checks fail because internally the `contrast_limits` are set to [60, 60] since its not monotonically increasing. They now fail with a `ValueError`.

**Comment by psobolewskiPhD** [comment]
> What is the expected behavior of `contrast_limits_range` is set outside of `contrast_limits`?

I think it should error. 
It shouldn't be possible to set the `range` outside the actual `limits`. Limits are the settings encompassed by the range (which determines the slider range).
Setting limits outside the range should be fine and reset the range accordingly.

Edit: why would the output on live of that be [60, 60]? Is it just pinning <60 to 60? That can't ever be the expected behavior can it?
Edit2: I guess this from @tlambert03 clarifies:
https://github.com/napari/napari/pull/4159#issuecomment-1053725079

**Comment by kcpevey** [comment]
>Setting limits outside the range should be fine and reset the range accordingly.

yes, this happens

As suggested (https://github.com/napari/napari/pull/4159#issuecomment-1053725079), I implemented an override of the contrast_limits if contrast_limits_range is outside of contrast_limits entirely. Otherwise it clips.

**Comment by psobolewskiPhD** [comment]
```
layer.contrast_limits = [20, 40]

layer.contrast_limits_range
Out[5]: [0, 100]

layer.contrast_limits_range = [60, 100]

layer.contrast_limits_range
Out[7]: [60, 100]

layer.contrast_limits
Out[8]: [60, 100]
```
Yea, to me the latest behavior is somewhat more intuitive—sort of?

If `range` is outside `limits` then `limits` are reset to `range`. So no more bizarre [60, 60].
On the surface of it this is reasonable...
But it's also wierd, because it means the image displayed is changed and there is no way to go back via the GUI, because the slider range is changed and doesn't cover the original range. 
Of course, such a change can only be made programatically, so maybe it doesn't matter?

But I still think that this is case where the user should get an error "contrast range doesn't encompass contrast limits" and be forced to change `contrast_limits` or `contrast_limits_range` explicitly, so they can't complain.

**Comment by brisvag** [comment]
> Is there any reason to keep the `range` on the napari/core side rather than the qt/gui side?

I still stand by this; I don't see why we would even *have* the concept of a "range" outside of the gui. This should only live on the qt side.

**Comment by psobolewskiPhD** [comment]
> > Is there any reason to keep the `range` on the napari/core side rather than the qt/gui side?
> 
> I still stand by this; I don't see why we would even _have_ the concept of a "range" outside of the gui. This should only live on the qt side.

This makes sense to me. The range should basically not be settable, it should be derived from the actual limits. Also, I think this touches on the need to have a reset button more accessible, as discussed in this related issue: https://github.com/napari/napari/issues/5127

**Comment by andy-sweet** [comment]
> > > Is there any reason to keep the `range` on the napari/core side rather than the qt/gui side?
> > 
> > 
> > I still stand by this; I don't see why we would even _have_ the concept of a "range" outside of the gui. This should only live on the qt side.
> 
> This makes sense to me. The range should basically not be settable, it should be derived from the actual limits. Also, I think this touches on the need to have a reset button more accessible, as discussed in this related issue: #5127

I think the problem is that the qt/gui API is not public. So if we shove it over there, then people developing against napari (e.g. plugin devs) have no stable way of setting up the GUI to present a sensible contrast limits slider, which is why I think was added in #837 (also see [relevant discussion on image.sc too](https://forum.image.sc/t/napari-lazy-loading-from-omero/32292)).

One real case (among others) where this might be relevant if I write a reader plugin that returns 12-bit sensor data that is stored in 16-bit uints. I really the contrast slider to have a range of [0, 2^12-1] (because values outside those limits are not possible), but I want to allow the user to adjust the actual contrast limits because their actual data range may be limited. What I don't want is napari to present a contrast slider with a range of [0, 2^16-1] and have a user ask why their image is black.

Without the explicit range, it has to be inferred (either from the dtype, which could be misleading as above) or by reading all the data (which may not feasible). Or we have to add more guessing magic (e.g. reading some subset of the data to set the initial limits).

Alternatively, we could consider public ways of setting this outside of the `Layer` API, but I don't know what that would look like or how we could easily keep it in sync with `contrast_limits`, which we clearly need some logic for.

**Comment by andy-sweet** [comment]
Also, if I understood correctly, this PR is trying to fix a buggy case where `contrast_limits` is set to be non-increasing.

I understand that `contrast_limits` and `contrast_limits_range` interact, but that seems like an independent issue that should directly block this PR, right?

**Comment by brisvag** [comment]
> I understand that `contrast_limits` and `contrast_limits_range` interact, but that seems like an independent issue that should directly block this PR, right?

Yes, defintely not blocking: better fix the bug than doing nothing. However, I brought this up because I felt like we're trying to solve a problem that shouldn't exist...

> Without the explicit range, it has to be inferred (either from the dtype, which could be misleading as above) or by reading all the data (which may not feasible). Or we have to add more guessing magic (e.g. reading some subset of the data to set the initial limits).

That's a good point and you gave some clear use cases... I don't see a cleaner way around this than what we currently have, so I guess I'll shut up about it :P

File: napari/layers/_tests/test_layer_attributes.py
Comment (andy-sweet) [inline napari/layers/_tests/test_layer_attributes.py:61]: Even though I don't expect any issues, I'd still add `np.random.seed(0)` as the first line of this test to make this deterministic.

File: napari/utils/validators.py
Comment (andy-sweet) [inline napari/utils/validators.py]: I'm not sure if this was part of the discussion around a previous version of this PR, but this feel like a premature refactor.

I think the only place this `validators` module is used is in the `intensity_mixin` module. Maybe it could be used elsewhere in the future, but right now I think we have defined 4 functions just to validate that `contrast_limits` and `contrast_limits_range` is a pair of increasing values.

Rather than expand the public API of napari and make the call sequence deeper, I'd suggest defining a private help function like

```python
def _is_increasing_pair(value: Any) -> bool:
    return isinstance(value, Sequence) and len(value) == 2 and value[0] < value[1]
```

And then use it in the setters like

```python
if not _is_increasing_pair(contrast_limits):
    raise ValueError(f'contrast_limits should be a pair of increasing values. Instead got {contrast_llimits}.')
```

We could define a function that raises instead, but then we have to pass the name of the attribute through, which hurts readability. You could also do a checked conversion to `Tuple` if we want to support generators, but that doesn't seem particularly important to me.

Despite the length of my comment here, this isn't blocking - I'm just relatively confused how we got here.

File: napari/layers/intensity_mixin.py
Comment (andy-sweet) [inline napari/layers/intensity_mixin.py:141]: This is a side-effect of a publicly settable property of some important layers, so this behavior should be documented in the docstring. Some of the side effects were not previously documented, but it would be good to improve things here.

I considered raising here instead, but that seems unnecessarily picky. I also considered warning, but if a user is setting this particular property I think they should know what they're doing.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
A few minor comments that should be addressed and a longer one that doesn't have to be. Happy to approve after that.

**Comment by andy-sweet** [comment]
> Yes, defintely not blocking: better fix the bug than doing nothing. However, I brought this up because I felt like we're trying to solve a problem that shouldn't exist...

Now I understand the interaction more, I'm more sympathetic. But I also think there are bigger issues to worry about this particular attribute, so still think this is worth getting in mostly as is.

File: napari/utils/validators.py
Comment (kcpevey) [reply napari/utils/validators.py]: This was indeed part of a debate in the previous version of this PR (https://github.com/napari/napari/pull/4159#discussion_r815508225)

**Comment by kcpevey** [comment]
@andy-sweet this is ready for another review.

File: napari/utils/validators.py
Comment (andy-sweet) [reply napari/utils/validators.py]: OK, given the amount of previous discussion here, I'll let this go in.

File: napari/utils/validators.py
Comment (andy-sweet) [inline napari/utils/validators.py]: As a public function, this should include at least one example and should follow the [numpydoc structure](https://numpydoc.readthedocs.io/en/latest/format.html) (e.g. [for `Raises`](https://numpydoc.readthedocs.io/en/latest/format.html#raises)).

Alternatively (and preferred in my opinion), we could declare this to be private (by adding an underscore prefix). And maybe better still just move it in the `intensity_mixin.py`, which is the only place it is used right now.

File: napari/utils/validators.py
Comment (andy-sweet) [inline napari/utils/validators.py]: As a public function, this should follow the [numpydoc style for examples](https://numpydoc.readthedocs.io/en/latest/format.html#examples).

Or consider making this private (see https://github.com/napari/napari/pull/5036/files#r985971453).

File: napari/layers/intensity_mixin.py
Comment (andy-sweet) [inline napari/layers/intensity_mixin.py]: Nit: override is unclear. Here's my suggestion.

```suggestion
        current contrast limits, the contrast limits will be reset to the new range.
```

Comment (kcpevey) [reply napari/layers/intensity_mixin.py]: good point.

File: napari/layers/intensity_mixin.py
Comment (andy-sweet) [inline napari/layers/intensity_mixin.py]: This makes it sound like the new value of `contrast_limits_range` is clipped based on the current value of `current_limits`. But isn't it the other way round?

Comment (kcpevey) [reply napari/layers/intensity_mixin.py]: I think I was just not right there. Below is the behavior. I reworded it.

```
layer.contrast_limits
Out[2]: [0, 255]

layer.contrast_limits_range
Out[3]: [0, 255]

layer.contrast_limits = [30, 90]

layer.contrast_limits_range
Out[5]: [0, 255]

layer.contrast_limits_range = [100, 150]

layer.contrast_limits
Out[7]: [100, 150]

layer.contrast_limits_range = [50, 120]

layer.contrast_limits
Out[9]: [100, 120]
```

Comment (psobolewskiPhD) [reply napari/layers/intensity_mixin.py]: Why doesn't that last command:
`layer.contrast_limits_range = [50, 120]`
make `layer.contrast_limits` be [50, 120] ?
(edit: I still think setting the `contrast_limits_range` like that should be an error, but if it's going to work, shouldn't it be consistent with the line `Out[7]`)
EDIT: for posterity, the behavior above is correct. Setting the range wider shouldn't change the limits. I got them swapped. Sorry.

Comment (kcpevey) [reply napari/layers/intensity_mixin.py]: Hmm. Yes, that is what we discussed. I must have had a regression at some point. I'll look into it.

Comment (kcpevey) [reply napari/layers/intensity_mixin.py]: @psobolewskiPhD I'm not sure how I got those results. I must have been on the wrong branch at the time. The tests are written in the way you have stated that you expect and they are passing (also, manually checking shows the correct expected behavior as well). 

Expected behavior:  
```
        If the requested contrast range limits are completely outside the
        current contrast limits, the range will be set as requested and the
        contrast limits will be reset to the new range.
```

I believe this PR is complete.

**Review by andy-sweet — APPROVED** [review]
Looks good to me. Will merge after 24 hours unless there are any objections.

File: napari/layers/intensity_mixin.py
Comment (psobolewskiPhD) [reply napari/layers/intensity_mixin.py]: > @psobolewskiPhD I'm not sure how I got those results. I must have been on the wrong branch at the time. The tests are written in the way you have stated that you expect and they are passing (also, manually checking shows the correct expected behavior as well).
> 
> Expected behavior:
> 
> ```
>         If the requested contrast range limits are completely outside the
>         current contrast limits, the range will be set as requested and the
>         contrast limits will be reset to the new range.
> ```
> 
> I believe this PR is complete.

Sorry, it was actually my bad. I got limits and limits_range confused.
```
layer.contrast_limits
Out[7]: [100, 150]

layer.contrast_limits_range = [50, 120]

layer.contrast_limits
Out[9]: [100, 120]
```
This is the proper behavior. Limits are set, the new range is wider than the bottom limit, so the limit is not reset, just the range. The upper range is more narrow, so that's reset.
And yea, tests pass. Sorry about the confusion!

**Pull request merged**
